Add telemetry functionality

francescolavra opened the pull request:
This PR implements sending at each boot an HTTPS POST request to the radar server. This functionality is enabled when the environment contains a RADAR_KEY entry, for example:
environment:(NANOS_VERSION:0.1.28 OPS_VERSION:0.1.13 RADAR_KEY:eyJhbGciOiJIUzI1NiIsIGrAJrvhT7bwnPEqYp3kofAMCVoK2CTJ6eQ).
The telemetry functionality is implemented in the radar klib, which depends on the TLS klib, which is being added as part of this commit.
The telemetry client verifies the server certificate when establishing a TLS session, using a PEM certificate hardcoded in the radar klib.

A unique boot identifier is generated at each boot and sent in the HTTPS request, along with the Nanos and Ops version numbers if found in the environment.

The TLS klib uses the Mbed TLS library, which is imported from a separate git repository.

wjhun reviewed (approved):
Everything looks good; only minor comments. It's nice to see the tls code map neatly into the connection handler interface.

francescolavra commented on klib/mbedtls_conf.h:
```diff
@@ -0,0 +1,73 @@
+typedef unsigned long uint64_t;
+
+typedef unsigned long size_t;
+typedef unsigned long uintptr_t;
```
I think it would be fine for other platforms too: as long as we are targeting 64-bit platforms only, these typedefs won't change, a long should be 64-bit on any 64-bit platform.

francescolavra commented on klib/mbedtls.c:
```diff
@@ -0,0 +1,301 @@
+    mbedtls_ctr_drbg_context ctr_drbg;
+    mbedtls_entropy_context entropy;
+    mbedtls_x509_crt cacert;
+    status (*direct_connect)(heap h, ip_addr_t *addr, u16 port, connection_handler ch);
```
I put in kern_funcs only those functions that are called directly by the mbedtls library (and so need to be in the mbed_tls_conf.h file which is included by the library code); the other function pointers are only used in our .c file and so I thought I could keep them private to this file.

francescolavra commented on klib/mbedtls.c:
```diff
@@ -0,0 +1,301 @@
+        }
+        break;
+    case tls_open:
+        b = little_stack_buffer(512);
```
512 is an arbitrary chunk size

francescolavra commented on klib/mbedtls.c:
```diff
@@ -0,0 +1,301 @@
+        }
+        if (conn->outgoing && (tls_out_internal(conn, 0) < 0))
+            /* An error happened and the connection has been closed. */
+            return STATUS_OK;
```
As things stand now, it's not possible to make an error status from a klib, because the timm* functions cannot be called: they are inline functions and contain a bunch of calls to kernel functions (such as intern, allocate_buffer, vbprintf, table_set), so if they are called from a klib they result in a linker error.

wjhun replied:
This could be covered with a kernel-side helper to make the status ... even a result with just the integer return val would be better than no status.

francescolavra commented on src/net/direct.c:
```diff
@@ -32,6 +32,34 @@ typedef struct qbuf {
+    direct d = allocate(h, sizeof(struct direct));
+    if (d == INVALID_ADDRESS) {
+        msg_err("allocation failed\n");
+        return 0;
```
OK, I removed the call to msg_err() and changed the return value (and the corresponding checks in the callers) to INVALID_ADDRESS.

francescolavra commented on src/net/direct.c:
```diff
@@ -156,17 +189,8 @@ static void direct_conn_err(void *z, err_t err)
 }
 
-static void direct_listen_err(void *z, err_t err)
+static direct_conn direct_conn_alloc(direct d, struct tcp_pcb *pcb)
```
The direct_conn struct (and the direct_conn_send closure) was being leaked already in the existing code. Now I embedded the closure in the direct_conn struct and added a direct_conn_dealloc(), which is called when a connection is closed by either end.

francescolavra commented on klib/mbedtls.c:
```diff
@@ -0,0 +1,301 @@
+    else if (conn->app_ch)
+        apply(conn->app_ch, 0); /* notify connection failure to application layer */
+    mbedtls_ssl_free(&conn->ssl);
+    deallocate(tls.h, conn, sizeof(*conn));
```
There was indeed a chance that conn->in is invoked after this deallocate, because the null buffer passed to conn->out to close the connection might be queued and processed asynchronously. Since there was no way for a direct connection client to know when a connection has been shut down, I modified the net direct code (and for consistency also the tls code) so that the receive handler is always invoked (with a null buffer) to notify a connection shutdown to the upper layer, regardless of which end of the connection initiated the shutdown. And the tls code now ensures that the tls_conn struct is deallocated after connection shutdown.

francescolavra commented on klib/mbedtls.c:
```diff
@@ -0,0 +1,301 @@
+    mbedtls_ctr_drbg_init(&tls.ctr_drbg);
+    mbedtls_entropy_init(&tls.entropy);
+    if (mbedtls_ctr_drbg_seed( &tls.ctr_drbg, mbedtls_entropy_func, &tls.entropy, 0, 0))
+        return KLIB_INIT_FAILED;
```
OK, rprintf is now exported to klibs, and the tls klib uses it to print error messages when configuration of the MbedTLS library fails.

francescolavra commented:
A couple of notable changes (beside what written in the other comments) in this updated PR:
- there was a missing allocation of klib virtual memory during klib loading, which, when loading multiple klibs, was causing the virtual address spaces of different klibs to overlap; fixed in https://github.com/nanovms/nanos/pull/1325/commits/f59178dda472e27106fbce090c38a02f1ae1d8db
- the telemetry code is now is its own klib instead of the main kernel; since it no longer receives a notification when the network interface is brought up, it implements a backoff timer to retry sending the boot message when an attempt fails
- in https://github.com/nanovms/nanos/pull/1325/commits/c062b9d32d186cfc71ef946435aa81e64c8fcfc6 there is some refactoring of existing kernel functions so that they can be exported to klibs, and some new functions have been added as well

wjhun commented:
> * there was a missing allocation of klib virtual memory during klib loading, which, when loading multiple klibs, was causing the virtual address spaces of different klibs to overlap; fixed in [f59178d](https://github.com/nanovms/nanos/commit/f59178dda472e27106fbce090c38a02f1ae1d8db)

I see what happened here. Originally, a 1GB virtual allocation was made from the virtual_huge heap. This is what is done elsewhere in the kernel when loading elfs, as there is address space to burn. So no need to reserve as that easily covers the span of any klib. But then, thinking klibs should be built with -mcmodel=kernel, I added the klib_heap, which is 4k page sized, to place klibs above the kernel image. But, in the process, I inadvertently changed the allocate from HUGE_PAGESIZE to PAGESIZE and called it a day. Oops.

The reserve is fine...I suppose it could break if there was fragmentation from adding / removing klibs in such a way that would allow the one page to be allocated but not enough space to reserve the whole thing. Maybe a more robust way would be to first walk the elf, figure the maximum span of loaded segments, and just allocate that much space with the need to reserve on map.

francescolavra commented:
I implemented allocation of klib virtual memory in a robust way as suggested, by using a new function called walk_elf() to calculate the total amount of memory needed before doing the allocation. And I added deallocation at klib unloading, which was missing.
Updated commit https://github.com/nanovms/nanos/pull/1325/commits/21a5e9cd417ff5a9dde1f263160e56c37038fc60

wjhun commented on src/runtime/buffer.c:
```diff
@@ -17,6 +17,24 @@ buffer allocate_buffer(heap h, bytes s)
 }
+KLIB_EXPORT(allocate_buffer);
+
+void buffer_write(buffer b, const void *source, bytes length)
```
Maybe it isn't significant here, but in general taking away inlined functions in data structure code could cause performance-related side-effects. What we could do for these is leave the inlines as-is, make kern_* wrappers like you did for register_timer, but add a macro like KLIB_EXPORT_RENAME to specify a string to use for the exported symbol name, so that they may be exported under their original names - buffer_write, buffer_read, etc.

francescolavra replied:
Ok, I added the KLIB_EXPORT_RENAME macro, and created kern_ wrappers for the inline buffer_read and buffer_write functions

wjhun reviewed (commented):
The new changes look good - just a few comments.

francescolavra commented on src/net/direct.c:
```diff
@@ -156,40 +204,55 @@ static void direct_conn_err(void *z, err_t err)
+
+static void direct_conn_dealloc(direct_conn dc)
+{
+    deallocate(dc->d->h, dc, sizeof(struct direct_conn));
```
OK, to remove the connection from conn_head I added `if (!client) list_delete(&dc->l);` in `direct_conn_closed()` (which is the only caller).

francescolavra commented on src/runtime/tuple.c:
```diff
@@ -52,6 +53,26 @@ void destruct_tuple(tuple t, boolean recursive)
+    vend(ap);
+    return s;
+}
+KLIB_EXPORT_RENAME(kern_timm, timm);
```
I created this kern_timm() function which is exported  as "timm" and is now used by the tls klib to create a status error.
